Create model compiler to build ModelNode from ModelNodeConfig.

tcsnfkx opened the pull request:
Population pool is not implemented.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)
<!-- Reviewable:end -->

tcsnfkx committed:
Create model compiler to build ModelNode from ModelNodeConfig.

Population pool is not implemented.

efoxepstein reviewed (changes requested):
> Reviewed 11 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 11 of 23 files reviewed, 7 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 24 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicJx3QB6zA-5j5uqXt:-MicJx3QB6zA-5j5uqXu:bhj748g) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L24)):*
> ```objc
> namespace wfa_virtual_people {
> 
> // Covnerts @config to CompiledNode recursively.
> ```

`Covnerts` typo?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicK461EJ8e1a2AuyUw:-MicK461EJ8e1a2AuyUx:bynwh6l) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L31)):*
> ```cpp
> 
> // This stores some information that will be used when building child nodes.
> struct CompilerContext {
> ```

Put all symbols not declared in the header into an anonymous namespace please
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKCsvENQqIzu_47Ps:-MicKCsvENQqIzu_47Pt:b-r14f3s) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L36)):*
> ```cpp
> 
> // Indicates whether the child node is selected by chance or condition.
> struct SelectBy {
> ```

Can you just use absl::variant? A struct isn't really appropriate (https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) because "The struct must not have invariants that imply relationships between different fields" but there's a relationship between `chance` and `condition`: only one will be set.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 55 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKlqR8M1FgrUZxyGn:-MicKlqR8M1FgrUZxyGo:bc1g06c) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L55)):*
> ```cpp
> };
> 
> absl::StatusOr<SelectBy> Compile(const ModelNodeConfig& config,
> ```

What's happening here?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 119 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicLRbI1blvE8iRim7H:-MicLRbI1blvE8iRim7I:bjimn86) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L119)):*
> ```cpp
> 
> // Create a BranchNode with single branch, which is a stop node.
> BranchNode CompileStop(const std::string& name) {
> ```

absl::string_view please
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicMOEL-mSpATSsyosi:-MicMOEL-mSpATSsyosj:b5kb03t) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L36)):*
> ```cpp
> 
> using AttributesUpdater = BranchNode::AttributesUpdater;
> using AttributesUpdaters = BranchNode::AttributesUpdaters;
> ```

Ditto for unnamed namespace
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/conditional_assignment, line 1 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicOc_u4sPB4zon27k8:-MicOc_u4sPB4zon27k9:b-3te41s) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/conditional_assignment#L1)):*
> ```
> condition { op: HAS name: "acting_demo.gender" }
> ```

Can you add a `.textproto` suffix to all of these files?


<!-- Sent from Reviewable.io -->

efoxepstein reviewed (commented):
> Reviewed 9 of 20 files at r1, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 7 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)



<!-- Sent from Reviewable.io -->

efoxepstein reviewed (commented):
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 7 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)



<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fix based on comments.

Fix typo.
Put unpublished symbols into anonymous namespace.
Make SelectBy a class rather than a struct.
Use absl::string_view instead of std::string.
Add ".textproto" to test data file names.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 7 of 23 files reviewed, 7 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 24 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicJx3QB6zA-5j5uqXt:-MieNQFH37lXFFOMRa0Z:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L24)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

`Covnerts` typo?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 31 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicK461EJ8e1a2AuyUw:-MieNSzQEXM4EAPJ4hXm:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L31)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Put all symbols not declared in the header into an anonymous namespace please
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKCsvENQqIzu_47Ps:-MieNU2B2u54wI8Su-ld:bvxxk0e) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L36)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Can you just use absl::variant? A struct isn't really appropriate (https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) because "The struct must not have invariants that imply relationships between different fields" but there's a relationship between `chance` and `condition`: only one will be set.
</blockquote></details>

Changed to class.

"An `absl::variant` should always hold a value of one of its alternative types (except in the "valueless by exception state")." https://github.com/abseil/abseil-cpp/blob/master/absl/types/variant.h

In our case, it is possible that neither chance nor condition is set.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 55 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKlqR8M1FgrUZxyGn:-MieJPpO57Evj0Xa6Ujd:b-19ailq) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L55)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

What's happening here?
</blockquote></details>

The implementation and comments are below.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 119 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicLRbI1blvE8iRim7H:-MieNOAx8G2jLskpHr3c:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L119)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

absl::string_view please
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicMOEL-mSpATSsyosi:-MieO6s85ysudzlquxLW:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L36)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Ditto for unnamed namespace
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/conditional_assignment, line 1 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicOc_u4sPB4zon27k8:-MieO8qQCQ6dnf_WOaEz:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/conditional_assignment#L1)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Can you add a `.textproto` suffix to all of these files?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

efoxepstein reviewed (changes requested):
> Reviewed 16 of 16 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 3 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKCsvENQqIzu_47Ps:-Mig9yf68kbxNgXSHXb9:b-h165xo) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L36)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

Changed to class.

"An `absl::variant` should always hold a value of one of its alternative types (except in the "valueless by exception state")." https://github.com/abseil/abseil-cpp/blob/master/absl/types/variant.h

In our case, it is possible that neither chance nor condition is set.
</blockquote></details>

You can have a third option in the variant indicating this....
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 55 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKlqR8M1FgrUZxyGn:-MigAJKrE7dgiEbZrm2g:b194ihf) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L55)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

The implementation and comments are below.
</blockquote></details>

Maybe document that this is forward-declared because of mutual recursion?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicMOEL-mSpATSsyosi:-MigAtqY8YucQZXh0aLi:b-suky4l) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L36)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can all the using aliases be within it?


<!-- Sent from Reviewable.io -->

wliue reviewed (changes requested):
> Reviewed 4 of 20 files at r1, 2 of 3 files at r3, 1 of 2 files at r4, 16 of 16 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 11 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @tcsnfkx)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicwcMe0Ttezvwrk4uQ:-MicwcMe0Ttezvwrk4uR:b4bk9u3) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L27)):*
> ```objc
> //
> // In the returned node, any child node is referenced by CompiledNode. No index
> // is set in any node.
> ```

This may be a little confusing to reader not familiar with the model proto. Do you feel it is important to note this?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 28 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihZmrW6jOFrzw1KvBt:-MihZmrX4pRxWROctIkc:b-1txc0j) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L28)):*
> ```objc
> // In the returned node, any child node is referenced by CompiledNode. No index
> // is set in any node.
> absl::StatusOr<CompiledNode> Compile(const ModelNodeConfig& config);
> ```

Maybe use a more descriptive name, e.g. CompileModel or CompileModelNode?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 80 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihaQXV8-2hgwg0K5AB:-MihaQXV8-2hgwg0K5AC:b-d3j8vb) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L80)):*
> ```
> Compile
> ```

Maybe use a more descriptive name?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 89 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihbANb3lGOtQZlqbyF:-MihbANb3lGOtQZlqbyG:b-36o8al) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L89)):*
> ```
>   bool has_chance = false;
>   bool has_condition = false;
> ```

It may be a littler cleaner if we first check that all select_by match, then build node. Could also check random_seed before building.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 146 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihkAs-B_UC1LZFYJeE:-MihkAs-B_UC1LZFYJeF:b-8c8b1q) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L146)):*
> ```cpp
> }
> 
> // Create a BranchNode with single branch, which is a stop node.
> ```

We might be able to optimize this a little bit by adding a StopNode child directly when there is no updater/multiplicity. But I think it is not critical because stop node is only used as leaf node in UpdateTree, and should only be a small percentage of all nodes.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.h, line 29 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-Mid-XpU4IgEllXV-zRj:-Mid-XpU4IgEllXV-zRk:bzb4xvk) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.h#L29)):*
> ```
> a
> ```

Nit: a -> an
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 43 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-Mid1G9HEsfL4c1kSdXt:-Mid1G9HEsfL4c1kSdXu:bh3hnvz) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L43)):*
> ```cpp
> 
> // Read textproto from file @path to @message.
> absl::Status ReadTextProtoFile(absl::string_view path,
> ```

Is there already a lib func for this? If not, maybe add this to common_cpp?
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 43 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MigitteEJrBekk4IpVv:-MigiisEE44ZehfUowZE:b443hs8) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L43)):*
> ```cpp
>       &config));
>   EXPECT_THAT(Compile(config).status(),
>               StatusIs(absl::StatusCode::kInvalidArgument, ""));
> ```

If intentionally don't want to match error message, could just use StatusIs(e).
But it is better to check the error message too, e.g. "Children". Same below.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fixes based on comments.

Update comments.
Rename Compile to CompileModel and CompileNode.
Move using statements into inner namespace.
Test the error message for error status.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 17 of 23 files reviewed, 10 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicwcMe0Ttezvwrk4uQ:-Mj7xwTR6bI3Ku0W_CVP:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L27)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

This may be a little confusing to reader not familiar with the model proto. Do you feel it is important to note this?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 28 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihZmrW6jOFrzw1KvBt:-Mj7xx4d-5t986Kae5Pu:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L28)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Maybe use a more descriptive name, e.g. CompileModel or CompileModelNode?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKCsvENQqIzu_47Ps:-Mj6rGyd8Mfff0Gj6Dx8:b-5k4dqw) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L36)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

You can have a third option in the variant indicating this....
</blockquote></details>

I think that would be less readable than a class.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 55 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKlqR8M1FgrUZxyGn:-Mj7y-Mp1Oqi94X631MR:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L55)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Maybe document that this is forward-declared because of mutual recursion?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 80 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihaQXV8-2hgwg0K5AB:-Mj7y92R1Rg-3JCrMrQO:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L80)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
> Compile
> ```

Maybe use a more descriptive name?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 89 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihbANb3lGOtQZlqbyF:-Mj6t68A0BYfN5UzICpv:b9c0gew) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L89)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
>   bool has_chance = false;
>   bool has_condition = false;
> ```

It may be a littler cleaner if we first check that all select_by match, then build node. Could also check random_seed before building.
</blockquote></details>

As the select_by is in the branches.nodes sub-messages, I think it is more readable to check after processing the sub-message. As we expect the execution to fail when any select_by mismatches or random_seed misses, the performance is not the first thing to consider here.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 146 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihkAs-B_UC1LZFYJeE:-Mj6uYSP7SUj1NmqGect:b8zzliv) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L146)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

We might be able to optimize this a little bit by adding a StopNode child directly when there is no updater/multiplicity. But I think it is not critical because stop node is only used as leaf node in UpdateTree, and should only be a small percentage of all nodes.
</blockquote></details>

Acknowledge.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicMOEL-mSpATSsyosi:-Mj7y4g6E1AOb7FDZbg-:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L36)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Can all the using aliases be within it?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc, line 43 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-Mid1G9HEsfL4c1kSdXt:-Mj7uuh04OZ6ambAwbWO:b-bjnnxe) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/specification_utils.cc#L43)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Is there already a lib func for this? If not, maybe add this to common_cpp?
</blockquote></details>

Will do in future PR.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 43 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MigitteEJrBekk4IpVv:-Mj7y6f09u0-HTOEvfaw:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L43)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

If intentionally don't want to match error message, could just use StatusIs(e).
But it is better to check the error message too, e.g. "Children". Same below.


</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

efoxepstein reviewed (changes requested):
> Reviewed 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 9 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 36 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicKCsvENQqIzu_47Ps:-MjAOUEn9Z-L_6Q3WviJ:b1z3zbe) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/a3e5c92a84381e968e98981eb151cf4c51d15f84/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L36)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

I think that would be less readable than a class.
</blockquote></details>

OK, sounds good.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 57 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjAP1tc2nERuyTs2gAG:-MjAP1tc2nERuyTs2gAH:b4usyl7) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/8cfbbad77f81247afe0b96c41ce8802e27cf66a4/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L57)):*
> ```cpp
>   }
> 
>   SelectBranchBy GetBy() { return by_; }
> ```

`GetBy() const {`? Ditto for other getters
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 97 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjAe9W30UBszo21BY9f:-MjAe9W30UBszo21BY9g:blgka8m) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/8cfbbad77f81247afe0b96c41ce8802e27cf66a4/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L97)):*
> ```cpp
>     ASSIGN_OR_RETURN(SelectBy select_by,
>                      CompileNode(config, context, *branch->mutable_node()));
>     switch (select_by.GetBy()) {
> ```

The style guide strongly prefers short functions: https://google.github.io/styleguide/cppguide.html#Write_Short_Functions and this is a little long.

Could it make sense to structure it like this?

```
switch (branches.nodes(0).select_by_case()) {
  case foo: return CompileChanceBranchNode(branches, ...);
  case bar: return CompileConditionBranchNode(branches, ...);
}
```


<!-- Sent from Reviewable.io -->

wliue reviewed (approved):
> Reviewed 5 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 7 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicwcMe0Ttezvwrk4uQ:-MjBaUn80ekTd-gCd0ES:btughel) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L27)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> // In the returned node, any child node is referenced by CompiledNode. No index
> // is set in any node. (In a CompiledNode, any child node can be referenced by
> // a CompiledNode sub-message, or an index which refers to another
> // CompiledNode.)
> ```
</details>


Maybe change to:

// In a CompiledNode, any child node can be referenced by
// a CompiledNode sub-message, or an index which refers to another
// CompiledNode. For the CompiledNode returned by this function,
// all child nodes are referenced by CompiledNode.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 89 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihbANb3lGOtQZlqbyF:-MjBbjec6ope8GfGDFVd:b5yg7r4) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L89)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

As the select_by is in the branches.nodes sub-messages, I think it is more readable to check after processing the sub-message. As we expect the execution to fail when any select_by mismatches or random_seed misses, the performance is not the first thing to consider here.
</blockquote></details>

The suggestion is not about the performance, but to separate the logic of error checking and node building. The current implementation is fine too, but it mixes the logic, and needs additional variables has_chance and has_condition for the check. Up to you though.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fixes based on comments.

Make the getter methods const in SelectBy.
Separate the compiling of branch nodes based on select_by.
Update comments.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 20 of 23 files reviewed, 7 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MicwcMe0Ttezvwrk4uQ:-MjCqawGD6dIqJcdO0NE:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/9791d1651eecb0b54992ca7a68317095c5404f11/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.h#L27)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
> // In the returned node, any child node is referenced by CompiledNode. No index
> // is set in any node. (In a CompiledNode, any child node can be referenced by
> // a CompiledNode sub-message, or an index which refers to another
> // CompiledNode.)
> ```

Maybe change to:

// In a CompiledNode, any child node can be referenced by
// a CompiledNode sub-message, or an index which refers to another
// CompiledNode. For the CompiledNode returned by this function,
// all child nodes are referenced by CompiledNode.

</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 89 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MihbANb3lGOtQZlqbyF:-MjChbw79GyAh6DijCOZ:b-w0myn7) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/6b71ed344ab9b9ad0c523e4107a1195460dcbcac/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L89)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

The suggestion is not about the performance, but to separate the logic of error checking and node building. The current implementation is fine too, but it mixes the logic, and needs additional variables has_chance and has_condition for the check. Up to you though.
</blockquote></details>

I think we still need `has_chance` and `has_condition` when the logic of error checking is separated out. Updated the handling of different select_by types.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 57 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjAP1tc2nERuyTs2gAG:-MjCqcek26rYAAh0i1Bz:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/8cfbbad77f81247afe0b96c41ce8802e27cf66a4/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L57)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

`GetBy() const {`? Ditto for other getters
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 97 at r6](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjAe9W30UBszo21BY9f:-MjCqZmb7TiWjgScl7DS:boyexb3) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/8cfbbad77f81247afe0b96c41ce8802e27cf66a4/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L97)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

The style guide strongly prefers short functions: https://google.github.io/styleguide/cppguide.html#Write_Short_Functions and this is a little long.

Could it make sense to structure it like this?

```
switch (branches.nodes(0).select_by_case()) {
  case foo: return CompileChanceBranchNode(branches, ...);
  case bar: return CompileConditionBranchNode(branches, ...);
}
```
</blockquote></details>

Done. Thanks!


<!-- Sent from Reviewable.io -->

efoxepstein reviewed (approved):
> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 5 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @wliue)



<!-- Sent from Reviewable.io -->

wliue reviewed (approved):
> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: all files reviewed, 3 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 126 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjHMSJBEUga8OJEaDCP:-MjHMSJBEUga8OJEaDCQ:b-vzfajt) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/502c20de6fa2ec6fc53a7b70976725e967bcbf04/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L126)):*
> ```
> conditon
> ```

Typo.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 139 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjHMcFj80iCvgHKN939:-MjHMcFk8FSjTNZm5llz:b-fiowoj) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/502c20de6fa2ec6fc53a7b70976725e967bcbf04/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L139)):*
> ```
>   switch (branches.nodes(0).select_by_case())
> ```

Check size first.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fixes based on comments.

Typo fix.
Check that the branches contain at least 1 node.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 21 of 23 files reviewed, 3 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @oliver-amzn, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 126 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjHMSJBEUga8OJEaDCP:-Mjkk4C9-y-KzcYGQS1m:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/502c20de6fa2ec6fc53a7b70976725e967bcbf04/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L126)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
> conditon
> ```

Typo.
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 139 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8#-MjHMcFj80iCvgHKN939:-Mjkk31R7R-q47fhXpTw:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/502c20de6fa2ec6fc53a7b70976725e967bcbf04/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L139)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
>   switch (branches.nodes(0).select_by_case())
> ```

Check size first.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Apply clang-format.

tcsnfkx reviewed (commented):
> Dismissed @wliue from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/8)* status: 21 of 23 files reviewed, all discussions resolved (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @my2457, @oliver-amzn, and @wliue)



<!-- Sent from Reviewable.io -->
